contracts/extensions/MultisigPermissions.sol

```diff
@@ -0,0 +1,640 @@
+          if (domainSkillId == colony.getDomain(1).skillId) {
+            permissionMask = ONLY_ROOT_ROLE_MASK;
+          } else {
+            permissionMask = ONLY_ARCHITECTURE_ROLE_MASK;
+          }
+        }
+      }
+
+      // slither-disable-next-line incorrect-equality
+      if (overallDomainSkillId == 0) {
+        overallDomainSkillId = domainSkillId;
+        overallPermissionMask = permissionMask;
+      } else {
+        // slither-disable-next-line incorrect-equality
+        require(
+          overallDomainSkillId == domainSkillId && overallPermissionMask == permissionMask,
+          "colony-multisig-invalid-motion"
+        );
+      }
+    }
+
+    return (overallDomainSkillId, overallPermissionMask);
+  }
+
+  // Copied from ColonyRoles.sol
+  function getUserRoles(address who, uint256 where) public view returns (bytes32) {
+    return userDomainRoles[who][where];
+  }
+
+  // Copied from ColonyRoles.sol
+  function setUserRoles(
+    uint256 _permissionDomainId,
+    uint256 _childSkillIndex,
+    address _user,
+    uint256 _domainId,
+    bytes32 _roles
+  ) public {
+    // This is not strictly necessary, since these roles are never used in subdomains
+    require(_roles & ROOT_ROLES == 0 || _domainId == 1, "multisig-bad-domain-for-role");
+
+    require(
+      validateDomainInheritance(_permissionDomainId, _childSkillIndex, _domainId),
+      "multisig-invalid-domain-inheritance"
+    );
+
+    // Allow this function to be called if the caller:
+    require(
+      // Has core root permissions OR
+      colony.hasUserRole(msgSender(), 1, ColonyDataTypes.ColonyRole.Root) ||
+        // Has core architecture, if we're using that permission in a child domain
+        (colony.hasUserRole(
+          msgSender(),
+          _permissionDomainId,
+          ColonyDataTypes.ColonyRole.Architecture
+        ) && (_permissionDomainId != _domainId)),
+      "multisig-caller-not-correct-permissions"
+    );
+
+    Domain memory domain = colony.getDomain(_domainId);
+
+    bytes32 existingRoles = getUserRoles(_user, _domainId);
+    bytes32 rolesChanged = _roles ^ existingRoles;
+    bytes32 roles = _roles;
+    bool setTo;
+
+    for (uint8 roleId; roleId < uint8(ColonyRole.NUMBER_OF_ROLES); roleId += 1) {
+      bool changed = uint256(rolesChanged) % 2 == 1;
+      if (changed) {
+        setTo = uint256(roles) % 2 == 1;
+        setUserRole(_user, _domainId, roleId, setTo);
+
+        if (setTo) {
+          domainSkillRoleCounts[domain.skillId][roleId] += 1;
+        } else {
+          domainSkillRoleCounts[domain.skillId][roleId] -= 1;
+        }
+
+        emit MultisigRoleSet(msgSender(), _user, _domainId, roleId, setTo);
+      }
+
+      roles >>= 1;
+      rolesChanged >>= 1;
+    }
+  }
+
+  // Internal functions
+
+  function getSig(bytes memory motion) internal pure returns (bytes4 sig) {
+    assembly {
+      sig := mload(add(motion, 0x20))
+    }
+  }
+
+  function validateDomainInheritance(
```

Comment:
A lot of external calls here. Might be worth making `validateDomainInheritence` public on the colony itself.